libcudacxx/include/cuda/__cmath/fast_modulo_division.h

```diff
@@ -0,0 +1,181 @@
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___CMATH_FAST_MODULO_DIVISION_H
+#define _LIBCUDACXX___CMATH_FAST_MODULO_DIVISION_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__cmath/ceil_div.h>
+#include <cuda/__cmath/ilog.h>
+#include <cuda/__cmath/pow2.h>
+#include <cuda/std/__type_traits/common_type.h>
+#include <cuda/std/__type_traits/is_integer.h>
+#include <cuda/std/__type_traits/is_signed.h>
+#include <cuda/std/__type_traits/make_nbit_int.h>
+#include <cuda/std/__type_traits/make_unsigned.h>
+#include <cuda/std/__type_traits/num_bits.h>
+#include <cuda/std/__utility/pair.h>
+#include <cuda/std/cstdint>
+#include <cuda/std/limits>
+
+#include <cuda/std/__cccl/prologue.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/***********************************************************************************************************************
+ * Extract higher bits after multiplication
+ **********************************************************************************************************************/
+
+template <typename _Tp, typename _Up>
+[[nodiscard]] _CCCL_API constexpr _CUDA_VSTD::common_type_t<_Tp, _Up> __multiply_extract_higher_bits(_Tp __x, _Up __y)
+{
+  using namespace _CUDA_VSTD;
+  static_assert(__cccl_is_integer_v<_Tp> && sizeof(_Tp) <= 8, "unsupported type");
+  static_assert(__cccl_is_integer_v<_Up> && sizeof(_Up) <= 8, "unsupported type");
+  if constexpr (is_signed_v<_Tp>)
+  {
+    _CCCL_ASSERT(__x >= 0, "__x must be non-negative");
+    _CCCL_ASSUME(__x >= 0);
+  }
+  if constexpr (is_signed_v<_Up>)
+  {
+    _CCCL_ASSERT(__y >= 0, "__y must be non-negative");
+    _CCCL_ASSUME(__y >= 0);
+  }
+  constexpr auto __mul_bits = ::cuda::next_power_of_two(__num_bits_v<_Tp> + __num_bits_v<_Up>);
+  using __larger_t          = __make_nbit_uint_t<__mul_bits>;
+  using __ret_t             = common_type_t<_Tp, _Up>;
+  auto __ret                = (static_cast<__larger_t>(__x) * __y) >> (__mul_bits / 2);
+  return static_cast<__ret_t>(__ret);
+}
+
+/***********************************************************************************************************************
+ * Fast Modulo/Division based on Precomputation
+ **********************************************************************************************************************/
+
+template <typename _Tp, bool _IsDivisorNotOne = false>
+class fast_mod_div
```

Comment:
I believe this class should be an implementation detail rather than something we expose publicly.

We already expose `cuda::std::div` and I am happy to make this faster, but I do not see the benefit of the additional exposed struct